apps/hash-api/src/storage/index.ts

```diff
@@ -92,36 +103,182 @@ export const setupStorageProviders = (
   return getUploadStorageProvider();
 };
 
+const isFileEntity = (entity: Entity): entity is Entity<FileProperties> =>
+  SYSTEM_TYPES.propertyType.fileStorageKey.metadata.recordId.baseUrl in
+    entity.properties &&
+  extractBaseUrl(fileUrlPropertyTypeUrl) in entity.properties;
```

Comment:
How we detect files in the frontend in is by seeing if the file's entity type ultimately inherits from `File`... the best way here would be to get the entity type and check its inheritance chain to see if it terminates in `File`. But that would be quite a costly step in front of every file download (including avatars).

I think this is fine, because the file storage key must match the entity id (since we extract the entity id from the key)... someone could create their own entity of a different type with this property, and ensure the entityId was contained in the right place in the key, but all that would happen is that the file download would fail since the object wouldn't exist in storage. I think it's fine to allow this failure to happen. They couldn't spoof access to someone else's file because the key must be consistent with the entity id.

Reply:
Understood, thanks for clarifying